ruff_lsp/server.py

```diff
@@ -736,47 +1023,116 @@ async def format_document(
+            original_source=document.source,
+            fixed_source=result.stdout.decode("utf-8"),
+            is_notebook_file=document.is_notebook_file(),
+        )
+        return WorkspaceEdit(
+            document_changes=[
+                _create_text_document_edit(document.uri, document.version, edits)
+            ]
+        )
+    elif document.kind is DocumentKind.Notebook:
+        notebook_document = LSP_SERVER.workspace.get_notebook_document(
+            notebook_uri=document.uri
+        )
+        if notebook_document is None:
+            log_warning(f"No notebook document found for {document.uri!r}")
+            return None
+
+        output_notebook_cells = cast(
+            Notebook, json.loads(result.stdout.decode("utf-8"))
+        )["cells"]
+        if len(output_notebook_cells) != len(notebook_document.cells):
+            log_warning(
+                f"Number of cells in the output notebook doesn't match the number of "
+                f"cells in the input notebook. Input: {len(notebook_document.cells)}, "
+                f"Output: {len(output_notebook_cells)}"
+            )
+            return None
+
+        cell_document_changes: list[TextDocumentEdit] = []
+        for cell_idx, cell in enumerate(notebook_document.cells):
+            if cell.kind is not NotebookCellKind.Code:
+                continue
+            cell_document = LSP_SERVER.workspace.get_text_document(cell.document)
+            edits = _fixed_source_to_edits(
+                original_source=cell_document.source,
+                fixed_source=output_notebook_cells[cell_idx]["source"],
+                is_notebook_file=True,
+            )
+            cell_document_changes.append(
+                _create_text_document_edit(
+                    cell_document.uri,
+                    cell_document.version,
+                    edits,
+                )
+            )
+
+        return WorkspaceEdit(document_changes=list(cell_document_changes))
+    else:
+        assert_never(document.kind)
 
-    if not result.stdout:
-        return []
 
-    new_source = _match_line_endings(document, result.stdout.decode("utf-8"))
+def _fixed_source_to_edits(
+    *, original_source: str, fixed_source: str | list[str], is_notebook_file: bool
+) -> list[TextEdit]:
+    """Converts the fixed source to a list of TextEdits.
+
+    If the fixed source is a list of strings, it is joined together to form a single
+    string with an assumption that the line endings are part of the strings itself.
+    """
+    if isinstance(fixed_source, list):
+        fixed_source = "".join(fixed_source)
+
+    new_source = _match_line_endings(original_source, fixed_source)
 
     # Skip last line ending in a notebook cell.
-    if document.uri.startswith("vscode-notebook-cell"):
+    if is_notebook_file:
         if new_source.endswith("\r\n"):
             new_source = new_source[:-2]
         elif new_source.endswith("\n"):
             new_source = new_source[:-1]
```

Comment:
This is more of a TODO for me to check if this is actually required now. Ruff adds and removes the newline at the end of every cell but I need to verify that behavior. I'm not sure what kind of example can be used here. I might have to check with a Ruff version when Jupyter Notebook support was not present.